lib/modules/manager/npm/post-update/pnpm.ts

```diff
@@ -26,9 +27,22 @@ export async function generateLockFile(
       (upgrade) =>
         upgrade.depType === 'packageManager' && upgrade.depName === 'pnpm'
     );
+    let pnpmEngine;
+    const rootPackageJson = upath.join(cwd, 'package.json');
+    const content = await readFile(rootPackageJson, 'utf8');
+    if (content) {
+      const packageJson: NpmPackage = JSON.parse(content);
+      const engines = packageJson?.engines;
+      pnpmEngine = engines['pnpm'];
+    }
+    if (pnpmUpdate?.newValue) {
+      pnpmEngine = pnpmUpdate.newValue;
+    }
     const pnpmToolConstraint: ToolConstraint = {
       toolName: 'pnpm',
-      constraint: pnpmUpdate ? pnpmUpdate.newValue : config.constraints?.pnpm,
+      constraint: config.constraints?.pnpm
```

Comment:
Is this being set in the extract function, or will it only come from config?

Reply:
i dont think we should remove the one we set at the extract, as it is meant for the Update pnpm engine, if i remove it, then we won't be able to update it, that's regression.

and it will not override the value as `pnpmUpdate.newValue` will be changed only if we are in the pnpm version upgrade, and its not related to our issue, since we are trying to update a different dependency in the child workspace, but the problem is running the binary install constraint.

what we are saying is:
If the customer asked for pnpm version in config, then run that version.
if the `package.json` has pnpm constraint that needs an update, then update it and run the updated one.
if not, then go to the root, and check if the `package.json` has `engine.pnpm` and take the constraint from there.

the things that are actually worrying me are :
1) what if the `engine.pnpm` was on the workspace child instead of the root? is that an option? im not familiar with pnpm
2) what if the root workspace wasn't at the root of the folder? then we wont find a `package.json` there.. should we do a lookup until we find the last `package.json`?

Reply:
> i dont think we should remove the one we set at the extract, as it is meant for the Update pnpm engine, if i remove it, then we won't be able to update it, that's regression.

I don't mean to remove it from the dependencies list, but instead to remove it from `constraints.pnpm`.

Reply:
@rarkins i see, so i should remove `constraints.pnpm = dep.currentValue` from extract because it overrides the `constraints.config`?
if i remove it, then it will not override the `constraints.config`, and will just take the `pnpm.update` from the current `package.json`.

is that right? or am i still missing something?

Reply:
@PhilipAbed yes, you're right! In other words we extract the contraints "lazily". I think in such cases we probably don't even need to have the extra logic to check "is there an update in this PR for pnpm itself?" because if there was, it would have been written to package.json already, and we would extract it.

Reply:
@rarkins 
So you are saying that the Constraints aren't needed in the the Extract+lookup section, but needed in the Post-Update section.


so i should remove this section from pnpm.ts :: generateLockFile ?
```
const pnpmUpdate = upgrades.find(
  (upgrade) =>
    upgrade.depType === 'packageManager' && upgrade.depName === 'pnpm'
);
```

so we are checking if pnpm has an Upgrade, and then setting the upgrade new value as a constraint? 
and we want to remove that.
i agree.

Now here's a case i still don't think is solved yet:
What if the `package-x/package.json` ( child workspace package json) had the Constraint( engine->pnpm:version X), but the constraint wasn't found on the Root `package.json`, is that a valid case?

Reply:
Yes.

If there is an `pnpm` update, the current or root `package.json` is already updated. So please refactor the code above to this:
1. Use `config.constraints?.pnpm` if defined
2. Extract from local `package.json`
3. Extract from root `package.json`

Reply:
Do we run the pnpm command on the local folder or the root?

Reply:
@rarkins only on the top folder to generate lock file

Reply:
![image](https://user-images.githubusercontent.com/42116482/161736900-ca8af6c3-72ad-4b5b-afb9-11a3c89c2a4d.png)

i believe its on a project level and not workspace :D

Reply:
i finished coding, considering its just on a root level.

so either Config or Root level engine. (config is priority)